ENIGMAsystem/SHELL/Universal_System/lua_table.h

```diff
@@ -61,59 +60,35 @@ T my_max(const T& lhs, const T& rhs) {
 }
 
 
-template <class T> struct lua_table
-{
-  // This is what kind of sparse container we'll be using
-  typedef std::map<size_t,T> lua_map_type;
-
-private:
-  //Stuff relating to dense storage.
-  T* dense;
-  size_t dn_reserve; //Total available units.
+template <class T> struct lua_table {
+# ifndef JUST_DEFINE_IT_RUN
+  typedef std::vector<T> dense_type;
+  typedef std::map<size_t,T> sparse_type;
+# else
+  typedef T dense_type[1];
+  typedef T sparse_type[1];
+# endif
 
-  //Sparse storage is easier.
-  lua_map_type sparse;
+ private:
+  dense_type dense;
+  sparse_type sparse;
 
-  //Our one item of book-keeping
-  size_t mx_size; //Actual max_element+1
+  // Our one item of book-keeping
+  size_t mx_size;  ///< highest accessed element + 1
 
 
-  void pick_up(const lua_table<T>& who)
-  {
-    dn_reserve = who.dn_reserve;
+  void pick_up(const lua_table<T>& who) {
     mx_size = who.mx_size;
-
-    //Free old memory?
-    if (dense) {
-      delete [] dense;
-      dense = NULL;
-    }
-
-    //Create a new dense chunk, copy over values.
-    dense = new T[dn_reserve];
-    for (size_t i=0; i<dn_reserve; i++) {
-      dense[i] = who.dense[i];
-    }
-
-    //Copy the sparse map
+    dense = who.dense;
     sparse = who.sparse;
   }
 
   void upsize(const size_t c)
   {
-    //Create a new dense section and copy over values; free old memory.
-    T* new_dense = new T[c];
-    if (dense) {
-      for (size_t i=0; i<dn_reserve; i++) {
-        new_dense[i] = dense[i];
-      }
-      delete [] dense;
-    }
-    dense = new_dense;
-    dn_reserve = c;
-
+    dense.resize(c);
+    mx_size = c;
```

Comment:
@JoshDreamland I had to resize `dense` to fix array_create_2d from segfaulting when fill was used to initialize it. I also had to set `mx_size` to fix `array_length_2d` which was returning 0. Please take a look and make sure it is correct.


```gml
var user_array_2d = array_create_2d(3, 5, 256);
for (i = 0; i < 5; i += 1) {
  for (j = 0; j < 3; j += 1) {
    show_message(user_array_2d[i,j]);
  }
}
```
This is the crash you get without my fix to upsize.
```
Thread 1 received signal SIGSEGV, Segmentation fault.
0x000000000047b3bc in variant::operator= (this=0x0, x=...)
    at Universal_System/var4.cpp:105
105     variant& variant::operator=(const variant x)            { rval.d = x.rval.d; if ((type = x.type) == tstr) sval = x.sval; return *this; }
(gdb) bt
#0  0x000000000047b3bc in variant::operator= (this=0x0, x=...)
    at Universal_System/var4.cpp:105
#1  0x000000000054882b in lua_table<variant>::fill (this=0x4905318,
    value=..., length=3) at Universal_System/lua_table.h:129
#2  0x0000000000533058 in var::var (this=0x29af090, value=..., length=3,
    height=5) at Universal_System/var4.h:202
#3  0x000000000045e666 in enigma_user::array_create_2d (length=3, height=5,
    value=...) at Universal_System/var_array.cpp:34
```